<dl class="dl-horizontal">
    <dt><span style="text-transform:capitalize" id="avg_coverage_type_{{ plot_id }}">Mean</span> coverage</dt>
    <dd><span id="avg_coverage_{{ plot_id }}"></span></dd>
</dl>
<div id="gene_plot_{{ plot_id }}">
    <span class="hidden-xs">
        <label for="display_coverage_metric_group_{{ plot_id }}">
            Display:
        </label>
        <span class="btn-group" data-toggle="buttons" id="display_coverage_metric_group_{{ plot_id }}">
            <button class="btn btn-primary active display_coverage_metric_buttons {{ plot_id }}"
                    id="display_coverage_overview_button" data-tooltip="Shows all exons">
                <input type="radio">Overview</input>
            </button>
            <button class="btn btn-primary display_coverage_metric_buttons {{ plot_id }}"
                    id="display_coverage_detail_button" data-tooltip="Zoomed in view (can scroll horizontally if needed)">
                <input type="radio">Detail</input>
            </button>
        </span>
        <label>
            <input type="checkbox" id="include_utrs_checkbox_{{ plot_id }}" value=""> Include UTRs in plot
        </label>
    </span>
    {% include 'coverage_selectors.html' %}
    <br/>
    <div id="gene_plot_container_container_{{ plot_id }}">
        <div id="loading_coverage_{{ plot_id }}">
            Loading coverage...
            <br/><br/>
        </div>
        <span id="gene_plot_axis_container_{{ plot_id }}" class="hidden-xs" style="width:50px; display: inline-block;"></span>
        <span id="gene_plot_container_{{ plot_id }}" style="overflow-x:scroll; display: inline-block;"></span>
        <a class="btn btn-success" id="coverage_plot_download_{{ plot_id }}" target="_blank" download="{{ plot_id }}_coverage">Save coverage plot</a>
        <a class="btn btn-success" id="exon_plot_download_{{ plot_id }}" target="_blank" download="{{ plot_id }}_exon">Save exon image</a>
        <a class="btn btn-success" id="cnv_plot_download_{{ plot_id }}" target="_blank" download="{{ plot_id }}_cnv">Save CNV image</a>
    </div>
</div>
<style>
  .chart div {
  font: 10px sans-serif;
  background-color: steelblue;
  text-align: right;
  padding: 3px;
  margin: 1px;
  color: white;
  }
</style>
<div class="chart"></div>
<div id="not_covered" style="display: none;">
    No coverage for this transcript.
</div>
<br/>
<script>

$(document).ready(function() {
    var plot_id = '{{ plot_id }}';
    var population_plot = plot_id == 'population';
    var gene_plot_container = '#gene_plot_container_' + plot_id;
    var include_utr_checkbox = '#include_utrs_checkbox_' + plot_id;
    var cnv_svg = 'cnv_svg_' + plot_id;
    var transcript = window.transcript;
    var coverage_stats = !population_plot ? window.coverage_stats : window.population_coverage_stats;
    var variants_in_transcript = window.variants_in_transcript;
    var cnvs = !population_plot ? window.cnvs : window.population_cnvs;
    if ($(window).width() < 768) {
        $(gene_plot_container).css('width', $(window).width() + "px");
    } else {
        $(gene_plot_container).css('width', $(window).width()*10/12 + "px");
    }
    precalc_coding_coordinates(transcript, coverage_stats, 'pos');
    precalc_coding_coordinates(transcript, variants_in_transcript, 'pos');

    // only show variants that have a coding coordinate
    variants_in_transcript = _.filter(variants_in_transcript, function(variant) {
        return variant.pos_coding != undefined;
    });

    // only show coding rects that have a coding coordinate
    coverage_stats = _.filter(coverage_stats, function(d) {
        return d.pos_coding != undefined;
    });
    $('#avg_coverage_' + plot_id).html(coverage_sum('mean', coverage_stats));
    $('#avg_coverage_x_' + plot_id).html(coverage_sum('30', coverage_stats) * 100 + '%');

    var new_data = create_new_data(coverage_stats, 'pos_coding');
    var new_data_skip_utr = create_new_data(coverage_stats, 'pos_coding_noutr');
    if (coverage_stats != null) {
        gene_chart(coverage_stats, new_data_skip_utr, variants_in_transcript, transcript, cnvs, gene_plot_container,
                cnv_svg);
        if (variants_in_transcript.length) {
            update_variants();
        }
        $('#loading_coverage_' + plot_id).hide();
    } else {
        $('#gene_plot_' + plot_id).hide();
        $('#not_covered_' + plot_id).show();
    }
    if (!population_plot)
        $('.coverage_subcat_selectors.' + plot_id).hide();
    // Change coverage plot
    $('.coverage_metric_buttons.' + plot_id).change(function () {
        var v = $(this).attr('id').replace('_covmet_button', '');
        if (v == 'covered') {
            $('#over_x_select_container_' + plot_id).show();
            v = $('#over_x_select_' + plot_id).val().replace('X', '');
        } else {
            if (population_plot)
                $('#average_select_container_' + plot_id).show();
            v = $("#average_select_" + plot_id).val();
        }
        var detail = $('.display_coverage_metric_buttons' + plot_id +'.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $(include_utr_checkbox).is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        change_coverage_chart(coverage_stats, plot_data, variants_in_transcript, transcript, detail, v, !include_utrs,
                gene_plot_container, cnv_svg, cnvs);
    });
    $('#over_x_select_' + plot_id).change(function () {
        var detail = $('.display_coverage_metric_buttons' + plot_id +'.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $(include_utr_checkbox).is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        $('#avg_coverage_type_x_' + plot_id).html($(this).val());
        $('#avg_coverage_x_' + plot_id).html(coverage_sum($(this).val().replace('X', ''), coverage_stats) * 100 + '%');
        change_coverage_chart(coverage_stats, plot_data, variants_in_transcript, transcript, detail, $(this).val().replace('X', ''),
                !include_utrs, gene_plot_container, cnv_svg, cnvs);
    });
    $('#average_select_' + plot_id).change(function () {
        var detail = $('.display_coverage_metric_buttons' + plot_id +'.active').attr('id').replace('display_coverage_', '').replace('_button', '');
        var include_utrs = $(include_utr_checkbox).is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        $('#avg_coverage_type_' + plot_id).html($(this).val());
        $('#avg_coverage_' + plot_id).html(coverage_sum($(this).val(), coverage_stats));
        change_coverage_chart(coverage_stats, plot_data, variants_in_transcript, transcript, detail, $(this).val(), !include_utrs,
                gene_plot_container, cnv_svg, cnvs);
    });
    $(include_utr_checkbox).change(function () {
        setTimeout(function() {
            var detail = $('.display_coverage_metric_buttons' + plot_id +'.active').attr('id').replace('display_coverage_', '').replace('_button', '');
            var v = $('.coverage_metric_buttons' + plot_id +'.active').attr('id').replace('_covmet_button', '');
            v = (v == 'covered') ? $('#over_x_select_' + plot_id).val().replace('X', '') : $("#average_select_" + plot_id).val();
            var include_utrs = $(include_utr_checkbox).is(':checked');
            var plot_data = include_utrs ? new_data : new_data_skip_utr;
            change_coverage_chart(coverage_stats, plot_data, variants_in_transcript, transcript, detail, v, !include_utrs,
                    gene_plot_container, cnv_svg, cnvs);
        }, 10);
    });

// Change exon diagram
    $('#inverted_checkbox_' + plot_id).change(function () {
        setTimeout(function () {
            var v = $('#inverted_checkbox_' + plot_id).is(':checked');
            change_track_chart_variant_size(variants_in_transcript, v, gene_plot_container);
        }, 10);
    });

    $('.consequence_display_buttons, #filtered_checkbox, .indel_display_buttons').change(function () {
        setTimeout(function() {
            update_variants();
            update_cnvs();
            refresh_links(plot_id, cnvs);
        }, 10);
    });

    $('.display_coverage_metric_buttons.' + plot_id).change(function () {
        var detail = $(this).attr('id').replace('display_coverage_', '').replace('_button', '');
        var v = $('.coverage_metric_buttons' + plot_id +'.active').attr('id').replace('_covmet_button', '');
        if (v == 'covered') {
            $('#over_x_select_container_' + plot_id).show();
            v = $('#over_x_select_' + plot_id).val().replace('X', '');
        } else {
            if (plot_id != 'sample')
                $('#average_select_container_' + plot_id).show();
            v = $("#average_select_" + plot_id).val();
        }
        var include_utrs = $(include_utr_checkbox).is(':checked');
        var plot_data = include_utrs ? new_data : new_data_skip_utr;
        change_coverage_chart(coverage_stats, plot_data, variants_in_transcript, transcript, detail, v, !include_utrs,
                gene_plot_container, cnv_svg, cnvs);
    });

{#    $("#coverage_plot_download").on('click', function() {#}
{#        window.location.href = set_plot_image('gene_plot_container', 0);#}
{#    });#}
{#    $("#exon_plot_download").on('click', function() {#}
{#        window.location.href = set_plot_image('gene_plot_container', 1);#}
{#    });#}
    //change CNVs svg
    $('#filtered_checkbox').change(function () {
        $("#" + cnv_svg).remove();
        var include_utrs = $('#include_utrs_checkbox').is(':checked');
        var detail = get_plot_detail();

        make_cnvs_svg(cnvs, window.transcript, detail, !include_utrs, gene_plot_container, cnv_svg);

    });

    refresh_links(plot_id, cnvs);
});
</script>
